zebra-consensus/src/primitives/halo2.rs

```diff
@@ -78,93 +58,102 @@ lazy_static::lazy_static! {
     pub static ref VERIFYING_KEY: ItemVerifyingKey = ItemVerifyingKey::build();
 }
 
-// === TEMPORARY BATCH HALO2 SUBSTITUTE ===
-//
-// These types are meant to be API compatible with the batch verification APIs
-// in bellman::groth16::batch, reddsa::batch, redjubjub::batch, and
-// ed25519-zebra::batch. Once Halo2 batch proof verification math and
-// implementation is available, this code can be replaced with that.
-
 /// A Halo2 verification item, used as the request type of the service.
 #[derive(Clone, Debug)]
 pub struct Item {
-    instances: Vec<orchard::circuit::Instance>,
-    proof: orchard::circuit::Proof,
+    bundle: orchard::bundle::Bundle<orchard::bundle::Authorized, Amount>,
+    sighash: SigHash,
+}
+
+impl ItemSize for Item {
+    fn item_size(&self) -> usize {
+        self.bundle.actions().len()
+    }
 }
 
 impl Item {
     /// Perform non-batched verification of this `Item`.
     ///
     /// This is useful (in combination with `Item::clone`) for implementing
     /// fallback logic when batch verification fails.
-    pub fn verify_single(&self, vk: &ItemVerifyingKey) -> Result<(), halo2::plonk::Error> {
-        self.proof.verify(vk, &self.instances[..])
+    pub fn verify_single(self, vk: &ItemVerifyingKey) -> bool {
+        let mut batch = BatchValidator::default();
+        batch.queue(self);
+        batch.validate(vk, thread_rng())
     }
 }
 
-/// A fake batch verifier that queues and verifies halo2 proofs.
-#[derive(Default)]
-pub struct BatchVerifier {
-    queue: Vec<Item>,
+trait QueueBatchVerify {
+    fn queue(&mut self, item: Item);
 }
 
-impl BatchVerifier {
-    /// Queues an item for fake batch verification.
-    pub fn queue(&mut self, item: Item) {
-        self.queue.push(item);
+impl QueueBatchVerify for BatchValidator {
+    fn queue(&mut self, Item { bundle, sighash }: Item) {
+        self.add_bundle(&bundle, sighash.0);
     }
+}
 
-    /// Verifies the current fake batch.
-    pub fn verify<R: RngCore + CryptoRng>(
-        self,
-        _rng: R,
-        vk: &ItemVerifyingKey,
-    ) -> Result<(), halo2::plonk::Error> {
-        for item in self.queue {
-            item.verify_single(vk)?;
-        }
+impl From<(&zebra_chain::orchard::ShieldedData, SigHash)> for Item {
+    fn from((shielded_data, sighash): (&zebra_chain::orchard::ShieldedData, SigHash)) -> Item {
+        let anchor = orchard::tree::Anchor::from_bytes(shielded_data.shared_anchor.into()).unwrap();
+        let authorization = orchard::bundle::Authorized::from_parts(
+            orchard::Proof::new(shielded_data.proof.0.clone()),
+            <[u8; 64]>::from(shielded_data.binding_sig).into(),
+        );
 
-        Ok(())
-    }
-}
+        let bundle = orchard::bundle::Bundle::from_parts(
```

Comment:
This PR should use the orchard bundle created when preparing to compute transaction sighashes instead of converting orchard data to the `orchard` crate `Bundle` type twice.